ansi/sixel/palette.go

```diff
@@ -0,0 +1,385 @@
+	length int
+	// sliceChannel is the axis that will be cut if this cube is cut in half
+	sliceChannel quantizationChannel
+	// score is a heuristic value: higher means this cube is more likely to be cut
+	score uint64
+	// pixelCount is how many pixels are contained in this cube
+	pixelCount uint64
+}
+
+// cubePriorityQueue is a heap used to sort quantizationCube objects in order to select the correct
+// one to cut next. Pop will remove the queue with the highest score
+type cubePriorityQueue []any
+
+func (p *cubePriorityQueue) Push(x any) {
+	*p = append(*p, x)
+}
+
+func (p *cubePriorityQueue) Pop() any {
+	popped := (*p)[len(*p)-1]
+	*p = (*p)[:len(*p)-1]
+	return popped
+}
+
+func (p *cubePriorityQueue) Len() int {
+	return len(*p)
+}
+
+func (p *cubePriorityQueue) Less(i, j int) bool {
+	left := (*p)[i].(quantizationCube)
+	right := (*p)[j].(quantizationCube)
+
+	// We want the largest channel variance
+	return left.score > right.score
+}
+
+func (p *cubePriorityQueue) Swap(i, j int) {
+	(*p)[i], (*p)[j] = (*p)[j], (*p)[i]
+}
+
+// createCube is used to initialize a new quantizationCube containing a region of the uniqueColors slice
+func (p *sixelPalette) createCube(uniqueColors []sixelColor, pixelCounts map[sixelColor]uint64, startIndex, bucketLength int) quantizationCube {
+	minRed, minGreen, minBlue, minAlpha := uint32(0xffff), uint32(0xffff), uint32(0xffff), uint32(0xffff)
+	maxRed, maxGreen, maxBlue, maxAlpha := uint32(0), uint32(0), uint32(0), uint32(0)
+	totalWeight := uint64(0)
+
+	// Figure out which channel has the greatest variance
+	for i := startIndex; i < startIndex+bucketLength; i++ {
+		r, g, b, a := uniqueColors[i].Red, uniqueColors[i].Green, uniqueColors[i].Blue, uniqueColors[i].Alpha
+		totalWeight += pixelCounts[uniqueColors[i]]
+
+		if r < minRed {
+			minRed = r
+		}
+		if r > maxRed {
+			maxRed = r
+		}
+		if g < minGreen {
+			minGreen = g
+		}
+		if g > maxGreen {
+			maxGreen = g
+		}
+		if b < minBlue {
+			minBlue = b
+		}
+		if b > maxBlue {
+			maxBlue = b
+		}
+		if a < minAlpha {
+			minAlpha = a
+		}
+		if a > maxAlpha {
+			maxAlpha = a
+		}
+	}
+
+	dRed := maxRed - minRed
+	dGreen := maxGreen - minGreen
+	dBlue := maxBlue - minBlue
+	dAlpha := maxAlpha - minAlpha
+
+	cube := quantizationCube{
+		startIndex: startIndex,
+		length:     bucketLength,
+		pixelCount: totalWeight,
+	}
+
+	if dRed >= dGreen && dRed >= dBlue && dRed >= dAlpha {
+		cube.sliceChannel = quantizationRed
+		cube.score = uint64(dRed)
+	} else if dGreen >= dBlue && dGreen >= dAlpha {
+		cube.sliceChannel = quantizationGreen
+		cube.score = uint64(dGreen)
+	} else if dBlue >= dAlpha {
+		cube.sliceChannel = quantizationBlue
+		cube.score = uint64(dBlue)
+	} else {
+		cube.sliceChannel = quantizationAlpha
+		cube.score = uint64(dAlpha)
+	}
+
+	// Boost the score of cubes with more pixels in them
+	cube.score *= totalWeight
+
+	return cube
+}
+
+type Ordered interface {

```

Comment:
We don't need to export these types. Probably we should move them to `utils.go`
```suggestion
type ordered interface {

```